fix: Constant fields for documents

gavindsouza opened the pull request:
### Changes:
- Mark `owner`, `creation` as constant fields for all documents.
- Make the field label bold in user message.
- Disallow setting standard fields like `creation`, `owner`. These fields will only be set on Document insert.
- Added 2 tests for the above changes.
- Refactor Assignment Rule
- Clear Document cache on Document updates/trashing
- Updated caniuse-lite package

Depends on https://github.com/frappe/frappe/pull/15521

gavindsouza committed:
fix: Make owner, creation fields constant for docs

gavindsouza committed:
style: Make fieldname bold in user message

gavindsouza committed:
test: Use db_set to change creation, status

gavindsouza committed:
test: Add test for owner, creation constants

surajshetty3416 commented:
We might need to have an exception for ToDo? Since in ToDo, the value for "Allocated To" (which points to "owner" field) can be changed after creation.

gavindsouza commented:
> We might need to have an exception for ToDo? Since in ToDo, the value for "Allocated To" (which points to "owner" field) can be changed after creation.

Should it be allowed though? I see that at this point, ToDo, Email Template, Chat Room, and Help Article doctypes have an owner field. But it was never allowed to edit these fields directly from the Desk IIRC. Neither does the controller make any changes to the document. I think it's safe to disallow changing it.

netchampfaris commented:
@surajshetty3416 @gavindsouza 

I think we should introduce another field `assigned_to` in ToDo. The `owner` field should fundamentally be not editable.

shadrak98 commented:
> I think we should introduce another field `assigned_to` in ToDo. The `owner` field should fundamentally be not editable.

@netchampfaris This was the first approach to this issue but then we saw a lot of dependency on the owner field.
https://github.com/frappe/frappe/pull/14086 you can check this PR.

surajshetty3416 commented:
closing due to inactivity... @gavindsouza re-open once the PR is ready

gavindsouza committed:
fix: Set owner & creation if new Document

via https://github.com/frappe/frappe/pull/14918/commits/a323d624eb3a59942aaac70ec95412140e70b7cb

gavindsouza committed:
test: Add test for disallowing setting tandard fields

Via https://github.com/frappe/frappe/pull/14918/commits/db008020f69408a1ce278a311645fb861b5a78d7

gavindsouza committed:
refactor(minor): Override clear_cache instead of separate hooks

Clear cached maps via Document.clear_cache

gavindsouza committed:
fix: Clear Document cache on rename, delete

gavindsouza committed:
refactor(minor): Assignment Rule

* Simplified and refactored logic
* Added type hints for where it gets really unbearable :')
* Minor perf improvements & better API usages

gavindsouza committed:
test(assign_rule): Setup and teardown for local testing

gavindsouza committed:
fix(ux): Auto Repeat's validation message has grammatical fixes

gavindsouza commented:
I could make exceptions for ToDo, Email Template, and Help Article for setting the owner field but that would break the system. Instead, I'd rather we change the DocTypes themselves.

---

Will raise (an)other PR(s) to make changes to their fields which would be a pre-requisite to getting this merged. @shadrak98's https://github.com/frappe/frappe/pull/14086 mitigates the issues with the ToDo DocType; I'll try building on top of that.

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into owner-unchange

gavindsouza commented:
@Mergifyio backport version-13-hotfix

gavindsouza committed:
fix: Remove extra Document.validate_owner validation

gavindsouza committed:
chore: Update caniuse-lite package
